### PR TITLE
feat: modify the request method in v4

### PR DESCRIPTION
**Pull request by ytvnr**
This policy has been migrated for V4 engine but remains backward compatible with V3 engine.

BREAKING CHANGE:
Requires APIM 3.20 minimum because it requires RxJava3.

When using a V4 Api definition, the request is directly modified to be overridden by the one from configuration in addition to modifying a context attribute.

**Issue**

https://gravitee.atlassian.net/browse/APIM-466

**Description**

The v4 implementation now modify directly the request object instead of setting an attribute in the context.
Some integration tests has been added to ensure backward compatibility.

---
**Gravitee.io Automatic Deployment**

🚀 A prerelease version of this package has been published on Gravitee's private artifactory, you can:
 - use it directly by updating your project with version: `2.0.0-apim-466-support-http-method-verride-SNAPSHOT`
 - download it from Artifactory [here](https://odbxikk7vo-artifactory.services.clever-cloud.com/gravitee-snapshots/io/gravitee/policy/gravitee-policy-override-http-method/2.0.0-apim-466-support-http-method-verride-SNAPSHOT/gravitee-policy-override-http-method-2.0.0-apim-466-support-http-method-verride-SNAPSHOT.zip)

### CODE REVIEW

File: src/main/java/io/gravitee/policy/overriderequestmethod/OverrideRequestMethodPolicy.java [hunk 1]
```diff
@@ -15,36 +15,34 @@
...
+        return "policy-override-request-method";
     }
 
-    @OnRequest
-    public void onRequest(Request request, Response response, ExecutionContext executionContext, PolicyChain policyChain) {
-        executionContext.setAttribute(ExecutionContext.ATTR_REQUEST_METHOD, configuration.getMethod());
-        policyChain.doNext(request, response);
+    @Override
+    public Completable onRequest(HttpExecutionContext ctx) {
+        return Completable.fromRunnable(() -> ctx.request().method(configuration.getMethod()));
```
Comment (guillaumelamirand) [inline src/main/java/io/gravitee/policy/overriderequestmethod/OverrideRequestMethodPolicy.java]: I would also set attribute in addition to setting the method on the request.

File: src/main/java/io/gravitee/policy/overriderequestmethod/OverrideRequestMethodPolicy.java [hunk 2]
```diff
@@ -15,36 +15,34 @@
...
      * @param configuration the associated configuration to the new policy instance
      */
     public OverrideRequestMethodPolicy(OverrideRequestMethodPolicyConfiguration configuration) {
-        this.configuration = configuration;
+        super(configuration);
+    }
+
+    @Override
+    public String id() {
+        return "policy-override-request-method";
```
Comment (guillaumelamirand) [inline src/main/java/io/gravitee/policy/overriderequestmethod/OverrideRequestMethodPolicy.java:47]:
```suggestion
        return "policy-override-http-method";
```
To respect github repository name.

Comment (ytvnr) [reply src/main/java/io/gravitee/policy/overriderequestmethod/OverrideRequestMethodPolicy.java:47]: I used the id from `plugin.properties` thinking we have to keep the same name, am I wrong ?

File: src/main/java/io/gravitee/policy/overriderequestmethod/OverrideRequestMethodPolicy.java (hunk 1 above)
Comment (ytvnr) [reply src/main/java/io/gravitee/policy/overriderequestmethod/OverrideRequestMethodPolicy.java]: ceinture bretelle alright

File: src/test/java/io/gravitee/policy/overriderequestmethod/OverrideRequestMethodPolicyIntegrationTest.java [hunk 3]
```diff
@@ -39,24 +39,32 @@
...
+    void shouldChangeMethod(String method, HttpClient client) throws InterruptedException {
         wiremock.stubFor(get("/endpoint").willReturn(ok()));
 
-        TestObserver<HttpResponse<Buffer>> obs = client.request(HttpMethod.valueOf(method), "/test").rxSend().test();
-
-        awaitTerminalEvent(obs);
-        obs
-            .assertComplete()
-            .assertValue(response -> {
+        client
```
Comment (guillaumelamirand) [inline src/test/java/io/gravitee/policy/overriderequestmethod/OverrideRequestMethodPolicyIntegrationTest.java:59]: Not big change but I would do .test() call before making the assertion. Something like that in pseudo code
client.rxRequest(HttpMethod.valueOf(method), "/test")
            .flatMap(request -> request.rxSend())
            .test()
            .assertValue(response -> {                 
                assertThat(response.statusCode()).isEqualTo(200);
            })
            .await()
            .assertComplete()
            .assertNoErrors();

Comment (ytvnr) [reply src/test/java/io/gravitee/policy/overriderequestmethod/OverrideRequestMethodPolicyIntegrationTest.java:59]: good catch

File: src/main/java/io/gravitee/policy/overriderequestmethod/OverrideRequestMethodPolicy.java [hunk 4]
```diff
@@ -16,35 +16,37 @@
...
     }
 
-    @OnRequest
-    public void onRequest(Request request, Response response, ExecutionContext executionContext, PolicyChain policyChain) {
-        executionContext.setAttribute(ExecutionContext.ATTR_REQUEST_METHOD, configuration.getMethod());
-        policyChain.doNext(request, response);
+    @Override
+    public Completable onRequest(HttpExecutionContext ctx) {
+        return Completable.fromRunnable(() -> {
+            ctx.setAttribute(ExecutionContext.ATTR_REQUEST_METHOD, configuration.getMethod());
```
Comment (jhaeyaert) [inline src/main/java/io/gravitee/policy/overriderequestmethod/OverrideRequestMethodPolicy.java]: Can you use the `ContextAttributes.ATTR_REQUEST_METHOD` constant instead ?

**Review by jhaeyaert — APPROVED** [review]
One comment about a constant usage. LGTM.

File: src/main/java/io/gravitee/policy/overriderequestmethod/OverrideRequestMethodPolicy.java (hunk 4 above)
Comment (ytvnr) [reply src/main/java/io/gravitee/policy/overriderequestmethod/OverrideRequestMethodPolicy.java]: sure

File: src/test/resources/apis/v4_definition/override-http-method.json [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      "entrypoints": [
+        {
+          "type": "http-proxy"
+        }
+      ]
+    }
+  ],
+  "endpointGroups": [
+    {
+      "name": "default",
```
Comment (jgiovaresco) [inline src/test/resources/apis/v4_definition/override-http-method.json]:
```suggestion
      "name": "default-group",
```

File: pom.xml [hunk 6]
```diff
@@ -30,13 +30,13 @@
...
+        <version>20.5</version>
     </parent>
 
     <properties>
-        <gravitee-apim-gateway-tests-sdk.version>3.18.0-SNAPSHOT</gravitee-apim-gateway-tests-sdk.version>
-        <gravitee-bom.version>2.5</gravitee-bom.version>
-        <gravitee-gateway-api.version>1.32.2</gravitee-gateway-api.version>
+        <gravitee-apim-gateway-tests-sdk.version>3.21.0-SNAPSHOT</gravitee-apim-gateway-tests-sdk.version>
+        <gravitee-bom.version>3.0.0</gravitee-bom.version>
+        <gravitee-gateway-api.version>2.1.0-apim-466-support-http-method-verride-SNAPSHOT</gravitee-gateway-api.version>
```
Comment (jgiovaresco) [inline pom.xml]:
```suggestion
        <gravitee-gateway-api.version>2.1.0-alpha.6</gravitee-gateway-api.version>
```

**Review by jgiovaresco — COMMENTED** [review]
Could you update the README accordingly 🙈

**Pull request merged**